### PR TITLE
Strenghten basic features of rectangular distributions

### PR DESCRIPTION
**Pull request by vasslitvinov**
* Ensure there is a complete implementation of some basic functionality for our rectangular distributions.

  Distributions: &nbsp; `BlockCycDist.chpl` &nbsp; `BlockDist.chpl` &nbsp; `CyclicDist.chpl` &nbsp; `DimensionalDist2D.chpl` &nbsp; `ReplicatedDist.chpl` &nbsp; `StencilDist.chpl`

  Methods: `dsiLow` `dsiHigh` `dsiAlignedLow` `dsiAlignedHigh` `dsiFirst` `dsiLast` `dsiStride` `dsiAlignment` `dsiNumIndices` `dsiDim` `dsiDim` `dsiDims` `dsiGetIndices` `dsiMember` `doiToString` `dsiSerialWrite` `dsiLocalSlice` `dsiIndexOrder` `dsiMyDist`

  These methods' implementations are identical for these distributions, redirecting to the `whole` domain. The only exception is that `StencilDom.dsiMember` returns true also if the index is within the fluff area. I placed them together under a "common redirects" comment in each distribution source file and removed existing scattered, partial implementations.

* Get `BlockCycDist` and `StencilDist` to compile for `idxType=int(32)`. This involved minor fixes to `ChapelDomain.chpl` and `ChapelRange.chpl`.

  This, plus some test fixes, allowed all of the robustness suite to compile for these two distributions. Here is the scorecard for the robustness suite:

| distribution | #failures before | #failures after |
|:---|---|---|
| DR --local | 0 | 0 |
| DR | 16 | 16 |
| block | 0 | 0 |
| blockcyclic | 103 | 32 |
| cyclic | 0 | 0 |
| dimensional | n/a | n/a |
| replicated | 0 | 0 |
| stencil | 86 | 18 |

  Testing is under gasnet except for "DR --local". In nightly testing we do not test DR under gasnet, blockcyclic, dimensional, stencil. Dimensional is not suitable for the robustness suite because it supports only 2-d domains.

  We are likely still ways away from getting these less-tested distributions to work with unsigned idxTypes -- because I have seen several places where we are using variables of `idxType` for stride-related values and, in StencilDist, for a value that can be outright negative.

* The above includes replacing `.size` with `.sizeAs(int)` to avoid a deprecation warning. We can replace these back wholesale once the deprecation warning goes away.

* Get Dimensional and Replicated to pass `test/domains/userAPI/test2Ddomain.chpl` adapted to apply these two domain maps, with the exception that localSubdomain() does not work for Dimensional on multiple locales. Replace ReplicatedDom.domRep and ReplicatedDom.redirectee() with ReplicatedDom.whole for simplicity and consistency.

* Beef up `distributions/robust/arithmetic/trivial` so it tests all the methods under the "ensure" bullet above, by adding a couple of tests and increasing the coverage in a couple of existing tests.

These new .good files are copies of existing ones:
&nbsp; from `test/domains/userAPI/test2Ddomain.good` :
`test/distributions/dm/testUserAPI.good`   (the original contents is replicated 4 times)
`test/distributions/replicated/testUserAPI.good`   (a symlink)
&nbsp; from `test/distributions/robust/arithmetic/basics/test_scan1.cyclic.good` :
`test/distributions/robust/arithmetic/basics/test_scan1.blockcyclic.good`
`test/distributions/robust/arithmetic/basics/test_scan1.stencil.good`

Testing: standard and gasnet paratest; distribution robustness suite (see the scorecard above).

### CODE REVIEW

File: modules/dists/StencilDist.chpl [hunk 1]
```diff
@@ -862,10 +849,26 @@ proc StencilDom.dsiBuildArray(type eltType, param initElts:bool) {
...
+proc StencilDom.dsiFirst         return whole.first;
+proc StencilDom.dsiLast          return whole.last;
+proc StencilDom.dsiStride        return whole.stride;
+proc StencilDom.dsiAlignment     return whole.alignment;
+proc StencilDom.dsiNumIndices    return whole.sizeAs(uint);
+proc StencilDom.dsiDim(d)        return whole.dim(d);
+proc StencilDom.dsiDim(param d)  return whole.dim(d);
+proc StencilDom.dsiDims()        return whole.dims();
+proc StencilDom.dsiGetIndices()  return whole.getIndices();
+//proc StencilDom.dsiMember(i)     return whole.contains(i);
```
Comment (bmcdonald3) [inline modules/dists/StencilDist.chpl:866]: Is this intentionally commented out?

File: modules/dists/DimensionalDist2D.chpl [hunk 2]
```diff
@@ -490,6 +490,30 @@ proc DimensionalDist2D.dsiClone(): _to_unmanaged(this.type) {
...
+
+proc DimensionalDom.dsiLocalSubdomain(loc: locale) {
+  import ChplConfig;
+  if ChplConfig.CHPL_COMM != "none" then
+    compilerError("DimensionalDist2D.dsiLocalSubdomain() is currently not implemented for non-local compilation");
+  return whole;
+  // for multilocale, start here:
+  // const (locIds, foundIt) = _CurrentLocaleToLocIDs(dist.targetLocales, loc);
+  // var result: whole.type;
+  // if !foundIt then return empty result;
```
Comment (bmcdonald3) [inline modules/dists/DimensionalDist2D.chpl:510]: Is this code intentionally commented out?

File: modules/dists/StencilDist.chpl (hunk 1 above)
Comment (vasslitvinov) [reply modules/dists/StencilDist.chpl:866]: Correct. In this and a couple of similar cases the domain map provides a different implementation for this feature that is elsewhere in the file.

File: modules/dists/DimensionalDist2D.chpl (hunk 2 above)
Comment (vasslitvinov) [reply modules/dists/DimensionalDist2D.chpl:510]: Yes. I started working on this method, however timed out, so left these to transcribe where I have gotten so far.

**Review by bmcdonald3 — APPROVED** [review]

**Pull request merged**